perf(ivy): avoid storing raw selectors in projectionDef

crisbeto opened the pull request:
Currently in Ivy we pass both the raw and parsed selectors to the `projectionDef` instruction, because the parsed selectors are used to match most nodes, whereas the raw ones are used to match against nodes with the `ngProjectAs` attribute. The raw selectors add a fair bit of code that won't be used in most cases, because `ngProjectAs` is somewhat rare.

These changes rework the compiler not to output the raw selectors in the `projectionDef`, but to parse the selector in `ngProjectAs` and to store it on the `TAttributes`. The logic for matching has also been changed so that it matches the pre-parsed `ngProjectAs` selector against the list of projection selectors.

petebacondarwin reviewed (commented):
I wonder if we should make `ngProjectAs` attributes special, in the sense that we create a new `AttributeMarker.ngProjectAs` flag. This would mean that our processing of "normal" attributes is not affected and fast as possible, which is the most common case.

crisbeto commented:
Introducing a flag would definitely make it more consistent, but I'm not sure whether it would help with processing since the `TAttributes` is still a `(string | number | CssSelector)[]` when we're iterating over it.

petebacondarwin commented on packages/core/src/render3/instructions/instructions.ts:
```diff
@@ -1920,7 +1920,7 @@ function generateInitialInputs(
 
     // If we hit any other attribute markers, we're done anyway. None of those are valid inputs.
-    if (typeof attrName === 'number') break;
+    if (typeof attrName !== 'string') break;
```
Do we guarantee that the `ngProjectAs` attribute will appear after all other attributes?

crisbeto replied:
Good point, we also want to break here if the `attrName === 'ngProjectAs'`.

mhevery replied:
I don't think we should have string comparison here as that is expensive. 

I do believe that `typeof attrName === 'number'` is faster than `typeof attrName !== 'string'` Why do we have to change it? Is this run on every instance or only for the initial creation? It looks like it is run for each instance, which means this change would have negative impact.

crisbeto replied:
The function is called `generateInitialInputs` so I think it's only being run on init.

WRT performance: we need to change it because the `attrName` is now a `string | number | CssSelector`. This was one of the concerns I was raising earlier. These are the two alternatives I have in mind:

Option one: Add another flag for `ngProjectAs` and use that to break the loop. That will allow us to change back the check to `typeof attrName === 'number'` and would be slightly less generated code.
Option two: Change the check to `attrName === 'ngProjectAs' || typeof attrName === 'number'`. I'm not quite sure how the performance would compare against option one.

gkalpak replied:
OOC, why is `typeof attrName === 'number'` faster than `typeof attrName !== 'string'`? Is it because of the extra negation that needs to happen for `!==`? Does this really have a noticeable impact?

mhevery reviewed (changes requested):
I am concerned about the changes we have made to the `generateInitialInputs` Could we discuss?

crisbeto commented:
I've reworked it to use a marker for `ngProjectAs`.

petebacondarwin reviewed (approved):
It would be good to see some tests that exercise the code with `projectAs` using classes and `not` too. Currently there only appears to be attribute based selectors. But I might have missed them.

crisbeto commented on packages/compiler-cli/test/compliance/r3_compiler_compliance_spec.ts:
```diff
@@ -1338,6 +1336,113 @@ describe('compiler compliance', () => {
+
+        const SimpleComponentDefinition = `
+          const $_c0$ = [[["", "title", ""]]];
+          const $_c1$ = ["[title]"];
```
The `c0` (parsed selectors) are still being passed into the `projectionDef` and the `c1` (raw selectors) are used in the `ComponentDef.ngContentSelectors`.

kara replied:
Seems like a comment might be good above the definition for this.

crisbeto replied:
Done.

crisbeto committed:
perf(ivy): avoid storing raw selectors in projectionDef

Currently in Ivy we pass both the raw and parsed selectors to the projectionDef instruction, because the parsed selectors are used to match most nodes, whereas the raw ones are used to match against nodes with the ngProjectAs attribute. The raw selectors add a fair bit of code that won't be used in most cases, because ngProjectAs is somewhat rare.

These changes rework the compiler not to output the raw selectors in the projectionDef, but to parse the selector in ngProjectAs and to store it on the TAttributes. The logic for matching has also been changed so that it matches the pre-parsed ngProjectAs selector against the list of projection selectors.

crisbeto commented:
Rebased and addressed the feedback @kara.

kara reviewed (approved):
LGTM

mhevery commented:
http://test/OCL:242709289:BASE:242709334:1554835459780:c2dfe02f

kara commented:
merge assistance: global approval

IgorMinar commented:
@crisbeto this PR now has a big conflict. can you please resolve it? thank you!

crisbeto committed:
fixup! perf(ivy): avoid storing raw selectors in projectionDef

crisbeto commented:
Conflicts have been resolved.

IgorMinar commented:
@crisbeto thank you! ❤️
